docs/contract-testing-howto.md

```diff
@@ -0,0 +1,73 @@
+
+The idea of contract testing is to gain more confidence that the whole system works together properly, but without needing to run all the services at the same time. Instead we can test each service in isolation using a 'contract' which describes the expectations on each interacting service.
+
+The definition of the contracts are provided by `govuk-content-schemas`.
+
+## Overview of publishing
+
+Publishing involves three systems:
+
+```
+                   (1)                                    (2)
+[publishing app] ------> [publishing api/content store] ------> [frontend app]
+
+(1) publishing representation
+(2) frontend representation
+
+```
+
+Note that the *frontend representation* differs slightly from the *publishing representation*. 
+ 
+The *contract* between the publishing application and the frontend application is defined by govuk-content-schemas. 
+
+This comprises three things:
+
+1. json-schema files which define the *publishing representation* for a given format
+2. a curated set of frontend examples of that format, which are validated against the schemas
+3. [a mechanism to convert from the 'publisher' schemas to the 'frontend' schemas](https://github.com/alphagov/govuk-content-schemas/blob/master/lib/govuk_content_schemas/frontend_schema_generator.rb), simulating the behaviour of the content store
+
+With those three parts we are able to verify that the examples against the schemas.
+
+This means that if the frontend works ok with the curated examples, and if the publishing tool produces output which matches the schema, then we can be quite confident that the frontends will work with the data produced by the publishing tool.
+
+## Adding a new format to govuk-content-schemas
+
+TODO
+
+## Setting up contract testing
+
+We have now set up contract testing between `whitehall` and `government-frontend`, so you can follow that as an example.
+ 
+The pieces required are:
+
+* [`jenkins-schema.sh` in government-frontend](https://github.com/alphagov/government-frontend/blob/master/jenkins-schema.sh) - runs the government frontend tests using the schema examples specified in the the `$SCHEMA_GIT_COMMIT` environment variable.
+* [`jenkins-schema.sh` in whitehall](https://github.com/alphagov/whitehall/blob/master/jenkins-schema.sh) - runs the whitehall publishing api presenter tests which take an edition model object and generate the JSON message which would be sent to the publishing API. This json is validated against the schemas.
+
+We then need to make sure that whenever a change is made to `govuk-
+content-schemas` we also run these two jobs.
+
+This is done using jenkins triggers on both the [master build](https://ci-new.alphagov.co.uk/job/govuk_content_schemas/) and the [branches build](https://ci-new.alphagov.co.uk/job/govuk_content_schemas_branches/) of `govuk-content-schemas`.
```

Comment:
I think it would be helpful if this was elaborating a bit more at a higher level. i.e. that when a change is made to the schema, the dependent "publishing" (e.g. whitehall) and "frontend" (e.g. government-frontend) builds are triggered to ensure that the changes have not broken the contracts. Maybe a step-by-step walk-through would help illustrate the interdependencies better?